File: src/security/LogTailer.cpp
#include "cachewraith/security/LogTailer.h"

#include "cachewraith/util/Logger.h"

#include <fstream>

namespace cachewraith {

LogTailer::LogTailer(std::filesystem::path path) : path_(std::move(path)) {}

const std::filesystem::path& LogTailer::path() const { return path_; }

std::vector<std::string> LogTailer::readNewLines() {
    std::error_code ec;
    if (!std::filesystem::exists(path_, ec)) {
        if (initialized_) {
            Logger::instance().warning("Log file is missing: " + path_.string());
        }
        return {};
    }

    const auto size = std::filesystem::file_size(path_, ec);
    if (ec) {
        Logger::instance().warning("Unable to stat log file: " + path_.string());
        return {};
    }

    if (!initialized_) {
        offset_ = size;
        last_size_ = size;
        initialized_ = true;
        return {};
    }

    if (size < last_size_) {
        Logger::instance().info("Log rotation detected for " + path_.string());
        offset_ = 0;
    }

    std::ifstream input(path_);
    if (!input) {
        Logger::instance().warning("Unable to read log file: " + path_.string());
        return {};
    }
    input.seekg(static_cast<std::streamoff>(offset_));

    std::vector<std::string> lines;
    std::string line;
    while (std::getline(input, line)) {
        lines.push_back(line);
    }
    offset_ = static_cast<std::uintmax_t>(input.tellg() >= 0 ? input.tellg() : static_cast<std::streampos>(size));
    last_size_ = size;
    return lines;
}

} // namespace cachewraith

File: agent-go/internal/logtailer/logtailer.go
package logtailer

import (
	"bufio"
	"log/slog"
	"os"
)

type Tailer struct {
	path        string
	offset      int64
	lastSize    int64
	initialized bool
}

func New(path string) *Tailer {
	return &Tailer{path: path}
}

func (t *Tailer) ReadNewLines() []string {
	info, err := os.Stat(t.path)
	if err != nil {
		if t.initialized {
			slog.Warn("log file is missing or unreadable", "path", t.path, "error", err)
		}
		return nil
	}

	size := info.Size()
	if !t.initialized {
		t.offset = size
		t.lastSize = size
		t.initialized = true
		return nil
	}
	if size < t.lastSize {
		slog.Info("log rotation detected", "path", t.path)
		t.offset = 0
	}

	file, err := os.Open(t.path)
	if err != nil {
		slog.Warn("unable to read log file", "path", t.path, "error", err)
		return nil
	}
	defer file.Close()

	if _, err := file.Seek(t.offset, 0); err != nil {
		slog.Warn("unable to seek log file", "path", t.path, "error", err)
		return nil
	}

	var lines []string
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		lines = append(lines, scanner.Text())
	}
	if pos, err := file.Seek(0, 1); err == nil {
		t.offset = pos
	} else {
		t.offset = size
	}
	t.lastSize = size
	return lines
}
